crates/ark/src/lsp/diagnostics.rs

```diff
@@ -780,13 +807,116 @@ fn recurse_call(
     let fun = fun.as_str();
 
     match fun {
-        // default case: recurse into each argument
-        _ => recurse_call_like_arguments_default(node, context, diagnostics)?,
+        "library" | "require" => {
+            // Track symbols exported by `library()` or `require()` calls
+            if let Err(err) = handle_package_attach_call(node, context) {
+                lsp::log_warn!("Can't handle attach call: {err:?}");
+            }
+        },
+        _ => {},
     };
 
+    // Continue with default recursion to handle any other arguments
+    recurse_call_like_arguments_default(node, context, diagnostics)?;
+
     ().ok()
 }
 
+fn handle_package_attach_call(node: Node, context: &mut DiagnosticContext) -> anyhow::Result<()> {
+    // Find the first argument (package name). Positionally for now, no attempt
+    // at argument matching whatsoever.
+    let Some(package_node) = node.arguments_values().nth(0) else {
+        return Err(anyhow::anyhow!("Can't unpack attached package argument"));
+    };
+
+    // Just bail if `character.only` is passed, even if it's actually `FALSE`.
+    // We'll do better when we have a more capable argument inspection
+    // infrastructure.
+    if let Some(_) = node
+        .arguments_names_as_string(context.contents)
+        .find(|n| n == "character.only")
+    {
+        return Ok(());
+    }
+
+    let package_name = package_node.get_identifier_or_string_text(context.contents)?;
+    let attach_pos = node.end_position();
+
+    let package = insert_package_exports(&package_name, attach_pos, context)?;
+
+    // Also attach packages from `Depends` field
+    let mut attach_dependencies = package.description.depends.clone();
```

Comment:
```suggestion
    for package_name in package.description.depends.iter() {
        insert_package_exports(package_name, attach_pos, context)?;
    }
```

I'd try and fully avoid the clone here, you just need a reference.

Down below, you'd just add a second loop of 

```suggestion
    for package_name in attach_field.iter() {
        insert_package_exports(package_name, attach_pos, context)?;
    }
```

I think that should clean things up fairly nicely in terms of clones / ownership of `Package` data and I think it reads a little nicer too